src/modules/commandrunner/src/lib/CommandRunner.cpp

```diff
@@ -1,465 +1,693 @@
+
+    while (!m_commandQueue.Empty())
+    {
+        std::shared_ptr<Command> command = m_commandQueue.Pop().lock();
+        if (nullptr != command)
+        {
+            command->Cancel();
+        }
+    }
+
+    // Push nullptr to signal the worker thread to exit
+    m_commandQueue.Push(std::weak_ptr<Command>());
 
     try
     {
-        commandWorkerThread.join();
+        if (m_workerThread.joinable())
+        {
+            m_workerThread.join();
+        }
+    }
+    catch (const std::exception& e)
+    {
+        OsConfigLogError(CommandRunnerLog::Get(), "Exception thrown while joining worker thread: %s", e.what());
+    }
+
+    Command::Status status = GetStatusToPersist();
+    if (!status.id.empty() && (0 != PersistCommandStatus(status)))
+    {
+        OsConfigLogError(CommandRunnerLog::Get(), "Failed to persist command status for client %s during shutdown", m_clientName.c_str());
     }
-    catch (const std::exception& e) {}
 }
 
-int CommandRunner::Run(CommandArguments command)
+int CommandRunner::GetInfo(const char* clientName, MMI_JSON_STRING* payload, int* payloadSizeBytes)
 {
     int status = MMI_OK;
-    if (commandArgumentsBuffer.empty() && commandWorkerThread.joinable())
+
+    if (nullptr == clientName)
     {
-        commandWorkerThread.join();
+        OsConfigLogError(CommandRunnerLog::Get(), "MmiGetInfo called with null clientName");
+        status = EINVAL;
     }
-
-    switch (command.action)
+    else if (nullptr == payload)
     {
-        case CommandRunner::Action::None:
-            OsConfigLogInfo(CommandRunnerLog::Get(), "No action to perform");
-            break;
-
-        case CommandRunner::Action::RefreshCommandStatus:
-            OsConfigLogInfo(CommandRunnerLog::Get(), "Refreshing commandId: %s", command.commandId.c_str());
-            status = SetCommandIdToRefresh(command.commandId);
-            break;
-
-        default:
-            AddCommandStatus(command.commandId, true);
-            UpdatePartialCommandStatus(command.commandId, 0, CommandState::Running);
+        OsConfigLogError(CommandRunnerLog::Get(), "MmiGetInfo called with null payload");
+        status = EINVAL;
+    }
+    else if (nullptr == payloadSizeBytes)
+    {
+        OsConfigLogError(CommandRunnerLog::Get(), "MmiGetInfo called with null payloadSizeBytes");
+        status = EINVAL;
+    }
+    else
+    {
+        try
+        {
+            std::size_t len = ARRAY_SIZE(info) - 1;
+            *payload = new (std::nothrow) char[len];
+            if (nullptr == *payload)
+            {
+                OsConfigLogError(CommandRunnerLog::Get(), "MmiGetInfo failed to allocate memory");
+                status = ENOMEM;
+            }
+            else
+            {
+                std::memcpy(*payload, info, len);
+                *payloadSizeBytes = len;
+            }
+        }
+        catch (const std::exception& e)

```

Comment:
>catch (const std::exception& e) [](http://example.com/codeflow?start=8&length=31)

Is this necessary? Regarding line 103. What other code can throw here?

Reply:
Good catch, I will update this later in the sample too.